cmd/cmd_test.go

```diff
@@ -20,20 +20,102 @@
 package cmd_test
 
 import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
 	"testing"
 
 	. "gopkg.in/check.v1"
 
 	"github.com/snapcore/snapd/cmd"
+	"github.com/snapcore/snapd/dirs"
 	"github.com/snapcore/snapd/release"
+	"github.com/snapcore/snapd/testutil"
 )
 
 func Test(t *testing.T) { TestingT(t) }
 
-type cmdSuite struct{}
+type cmdSuite struct {
+	restoreExec   func()
+	execCalled    int
+	lastExecArgv0 string
+	lastExecArgv  []string
+	lastExecEnvv  []string
+	newCore       string
+	oldCore       string
+}
 
 var _ = Suite(&cmdSuite{})
 
+func (s *cmdSuite) SetUpTest(c *C) {
+	s.restoreExec = cmd.MockSyscallExec(s.syscallExec)
+	s.execCalled = 0
+	s.lastExecArgv0 = ""
+	s.lastExecArgv = nil
+	s.lastExecEnvv = nil
+	d := c.MkDir()
+	s.newCore = filepath.Join(d, "new")
+	s.oldCore = filepath.Join(d, "old")
+}
+
+func (s *cmdSuite) TearDownTest(c *C) {
+	s.restoreExec()
+}
+
+func (s *cmdSuite) syscallExec(argv0 string, argv []string, envv []string) (err error) {
+	s.execCalled++
+	s.lastExecArgv0 = argv0
+	s.lastExecArgv = argv
+	s.lastExecEnvv = envv
+	return fmt.Errorf(">exec of %q in tests<", argv0)
+}
+
+func (s *cmdSuite) fakeCoreVersion(c *C, coreDir, version string) {
+	p := filepath.Join(coreDir, "/usr/lib/snapd")
+	c.Assert(os.MkdirAll(p, 0755), IsNil)
+	c.Assert(ioutil.WriteFile(filepath.Join(p, "info"), []byte("VERSION="+version), 0644), IsNil)
+}
+
+func (s *cmdSuite) fakeInternalTool(c *C, coreDir, toolName string) string {
+	s.fakeCoreVersion(c, coreDir, "9999")
+	p := filepath.Join(coreDir, "/usr/lib/snapd", toolName)
+	c.Assert(ioutil.WriteFile(p, nil, 0755), IsNil)
+
+	return p
+}
+
+func (s *cmdSuite) mockReExecingEnv() func() {
+	restore := []func(){
+		release.MockOnClassic(true),
+		release.MockReleaseInfo(&release.OS{ID: "ubuntu"}),
+		cmd.MockCorePaths(s.oldCore, s.newCore),
+		cmd.MockVersion("2"),
+	}
+
+	return func() {
+		for i := len(restore) - 1; i >= 0; i-- {
+			restore[i]()
+		}
+	}
+}
+
+func (s *cmdSuite) mockReExecFor(c *C, coreDir, toolName string) func() {
+	selfExe := filepath.Join(coreDir, "proc-self-exe")
```

Comment:
this still confuses me conceptually,  as I said I would have captured the top MkDir and used s.tmpdir or s.fakeroot